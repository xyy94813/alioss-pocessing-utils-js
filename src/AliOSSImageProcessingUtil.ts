/*
 * File Created: Wednesday, 14th November 2018 3:15:26 pm
 * Author: xyy94813 (xyy94813@sina.com)
 * -----
 * Last Modified: Monday, 19th November 2018 6:37:47 pm
 * Modified By: xyy94813 (xyy94813@sina.com>)
 */
import AliOSSProcessingUtil from './AliOSSProcessingUtil';

export enum ImageOperation {
  RESIZE = 'resize',
  CIRCLE = 'circle',
  CROP = 'crop',
  INDEXCROP = 'indexcrop',
  ROUNDED_CORNERS = 'rounded-corners',
  AUTO_ORIENT = 'auto-orient',
  ROTATE = 'rotate',
  BLUR = 'blur',
  BRIGHT = 'bright',
  CONTRAST = 'contrast',
  SHARPEN = 'sharpen',
  FORMAT = 'format ',
  INTERLACE = 'interlace',
  QUALITY = 'quality',
  WATERMARK = 'watermark',
  AVERAGE_HUE = 'average-hue',
  INFO = 'info',
}

export enum ImageResizeMode {
  LFIT = 'lfit', // proportional scaling. It refers to the maximum image that is limited in the rectangle of the specified w and h.
  MFIT = 'mfit', // proportional scaling. It refers to the minimum image extending out of the rectangle of the specified w and h.
  FILL = 'fill', // fixed width and height. It refers to the cropped and centered minimum image extending out of the rectangle of the specified w and h.
  PAD = 'pad', // fixed width and height, scaling down and filling.
  FIXED = 'fixed', // fixed width and height, enforced scaling down.
}

export interface IGetAliOSSImageResizeAPIOptions {
  mode?: ImageResizeMode; // one of lfit、mfit、fill、pad、fixed
  width?: number; // Specify the target width.
  height?: number; // Specify the target height.
  longSide?: number; // Specify the longer side of the target.
  shorterSide?: number; // Specify the shorter side of the target.
  limit?: number; // 0/1. The default value is 1
  color?: string; // When you set the scaling mode as pad (scaling down and filling), you can select the filling color (The default is white).
  // Filling format of parameters: use hexadecimal color codes, for example 00FF00 (green).
  percent?: number; // Percentage. If it is smaller than 100, it means to scale down; if it is bigger than 100, it means to scale up.
}

export interface IGetAliOSSImageCircleAPIOptions {
  radius: number; // Radius of the circular area of the image
}

export enum ImageCropOrigin {
  NW = 'nw',
  NORTH = 'north',
  NE = 'ne',
  WEST = 'west',
  CENTER = 'center',
  EAST = 'east',
  SW = 'sw',
  SOUTH = 'south',
  SE = 'se',
}

export interface IGetAliOSSImageCropAPIOptions {
  width: number; // Width of the cropped area
  height: number; // Height of the cropped area
  xAxis?: number; // 	X-axis of the crop starting point (the origin is located in the upper-left corner by default)
  yAxis?: number;
  orign?: ImageCropOrigin;
}

export interface IGetAliOSSImageIndexcropAPIOptions {
  i: number;
  x?: number; // Length of each image partition during horizontal cutting. Either the x or y parameter must be used.
  y?: number; // Length of each image partition during vertical cutting. Either the x or y parameter must be used.
}

export interface IGetAliOSSImageRoundedCornersAPIOptions {
  radius: number; // Radius of the circular area of the image
}

export interface IGetAliOSSImageAutoOrientAPIOptions {
  autoOrient: boolean;
}

export interface IGetAliOSSImageRotateAPIOptions {
  rotate: number;
}

interface IAliOSSImageProcessingUtil {
  getAliOSSImageProcessingAPI(operation: ImageOperation, vals: any): string;
  getAliOSSImageResizeAPI(options: IGetAliOSSImageResizeAPIOptions): string;
  getAliOSSImageCircleAPI(options: IGetAliOSSImageCircleAPIOptions): string;
  getAliOSSImageCropAPI(options: IGetAliOSSImageCropAPIOptions): string;
  getAliOSSImageIndexcropAPI(
    options: IGetAliOSSImageIndexcropAPIOptions
  ): string;
  getAliOSSImageRoundedCornersAPI(
    options: IGetAliOSSImageRoundedCornersAPIOptions
  ): string;
  getAliOSSImageAutoOrientAPI(
    options: IGetAliOSSImageAutoOrientAPIOptions
  ): string;
  getAliOSSImageRotateAPI(options: IGetAliOSSImageRotateAPIOptions): string;
}

class AliOSSImageProcessingUtil extends AliOSSProcessingUtil
  implements IAliOSSImageProcessingUtil {
  public getAliOSSImageProcessingAPI = (
    operation: ImageOperation,
    vals: any
  ) => {
    return this.getAliOSSProcessingAPI('image', operation, vals);
  };
  public getAliOSSImageResizeAPI = (
    options: IGetAliOSSImageResizeAPIOptions
  ) => {
    const {
      mode,
      width,
      height,
      longSide,
      shorterSide,
      limit,
      percent,
    } = options;
    return this.getAliOSSImageProcessingAPI(ImageOperation.RESIZE, {
      h: height,
      l: longSide,
      limit,
      m: mode,
      percent,
      s: shorterSide,
      w: width,
    });
  };
  public getAliOSSImageCircleAPI = (
    options: IGetAliOSSImageCircleAPIOptions
  ) => {
    return this.getAliOSSImageProcessingAPI(ImageOperation.CIRCLE, {
      r: options.radius,
    });
  };
  public getAliOSSImageCropAPI = (options: IGetAliOSSImageCropAPIOptions) => {
    const { width, height, xAxis, yAxis, orign } = options;
    return this.getAliOSSImageProcessingAPI(ImageOperation.CROP, {
      g: orign,
      h: height,
      w: width,
      x: xAxis,
      y: yAxis,
    });
  };
  public getAliOSSImageIndexcropAPI = (
    options: IGetAliOSSImageIndexcropAPIOptions
  ) => {
    return this.getAliOSSImageProcessingAPI(ImageOperation.INDEXCROP, options);
  };

  public getAliOSSImageRoundedCornersAPI = (
    options: IGetAliOSSImageRoundedCornersAPIOptions
  ) => {
    return this.getAliOSSImageProcessingAPI(ImageOperation.ROUNDED_CORNERS, {
      r: options.radius,
    });
  };
  public getAliOSSImageAutoOrientAPI = (
    options: IGetAliOSSImageAutoOrientAPIOptions
  ) => {
    return this.getAliOSSImageProcessingAPI(ImageOperation.AUTO_ORIENT, {
      value: options.autoOrient ? 1 : 0,
    });
  };
  public getAliOSSImageRotateAPI = (
    options: IGetAliOSSImageRotateAPIOptions
  ) => {
    return this.getAliOSSImageProcessingAPI(ImageOperation.ROTATE, {
      value: options.rotate,
    });
  };
}

export default AliOSSImageProcessingUtil;
