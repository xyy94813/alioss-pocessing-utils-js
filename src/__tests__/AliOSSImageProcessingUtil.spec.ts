/*
 * File Created: Wednesday, 14th November 2018 3:26:13 pm
 * Author: xyy94813 (xyy94813@sina.com)
 * -----
 * Last Modified: Monday, 8th April 2019 3:56:32 pm
 * Modified By: xyy94813 (xyy94813@sina.com>)
 */
import AliOSSImageProcessingUtil, {
  ImageOperation,
  ImageResizeMode,
} from '../AliOSSImageProcessingUtil';

const originUrl = 'https://image-demo.oss-cn-hangzhou.aliyuncs.com/example.jpg';
const util = new AliOSSImageProcessingUtil(originUrl);

describe('AliOSSImageProcessingUtil', () => {
  describe('getAliOSSImageResizeAPI', () => {
    describe('with mode', () => {
      it('mode is fill', () => {
        expect(
          util.getAliOSSImageResizeAPI({
            mode: ImageResizeMode.FILL,
            width: 200,
          })
        ).toMatchSnapshot();
      });
      it('mode is fixed', () => {
        expect(
          util.getAliOSSImageResizeAPI({
            mode: ImageResizeMode.FIXED,
            width: 200,
          })
        ).toMatchSnapshot();
      });
      it('mode is lfit', () => {
        expect(
          util.getAliOSSImageResizeAPI({
            mode: ImageResizeMode.LFIT,
            width: 200,
          })
        ).toMatchSnapshot();
      });
      it('mode is mfit', () => {
        expect(
          util.getAliOSSImageResizeAPI({
            mode: ImageResizeMode.MFIT,
            width: 200,
          })
        ).toMatchSnapshot();
      });
      it('mode is pad', () => {
        expect(
          util.getAliOSSImageResizeAPI({
            mode: ImageResizeMode.PAD,
            width: 200,
          })
        ).toMatchSnapshot();
      });
    });
  });
  describe('getAliOSSImageCircleAPI', () => {
    it('radius is normal', () => {
      expect(
        util.getAliOSSImageCircleAPI({
          radius: 100,
        })
      ).toMatchSnapshot();
    });
  });
  describe('getAliOSSImageCropAPI', () => {
    it('set width and height', () => {
      expect(
        util.getAliOSSImageCropAPI({
          height: 100,
          width: 100,
        })
      ).toMatchSnapshot();
    });
  });
  describe('getAliOSSImageIndexcropAPI', () => {
    it('x is 100 and i is 0', () => {
      expect(
        util.getAliOSSImageIndexcropAPI({
          i: 100,
          x: 100,
        })
      ).toMatchSnapshot();
    });
  });
  describe('getAliOSSImageRoundedCornersAPI', () => {
    it('set radius', () => {
      expect(
        util.getAliOSSImageRoundedCornersAPI({
          radius: 30,
        })
      ).toMatchSnapshot();
    });
  });
  describe('getAliOSSImageAutoOrientAPI', () => {
    it('auto origent', () => {
      expect(
        util.getAliOSSImageAutoOrientAPI({
          autoOrient: true,
        })
      ).toMatchSnapshot();
    });
    it('not auto origent', () => {
      expect(
        util.getAliOSSImageAutoOrientAPI({
          autoOrient: false,
        })
      ).toMatchSnapshot();
    });
  });
  describe('getAliOSSImageRotateAPI', () => {
    it('set rotate', () => {
      expect(
        util.getAliOSSImageRotateAPI({
          rotate: 30,
        })
      ).toMatchSnapshot();
    });
  });
  describe('getAliOSSImageProcessingAPI', () => {
    it('mutiple operation', () => {
      expect(
        util.getAliOSSImageProcessingAPI([
          {
            operation: ImageOperation.RESIZE,
            vals: {
              h: 200,
              w: 200,
            },
          },
          {
            operation: ImageOperation.ROTATE,
            vals: {
              value: 90,
            },
          },
        ])
      ).toMatchSnapshot();
    });
  });
  describe('origin url with query', () => {
    it('without param', () => {
      expect(
        new AliOSSImageProcessingUtil(
          `${originUrl}?`
        ).getAliOSSImageAutoOrientAPI({
          autoOrient: false,
        })
      ).toMatchSnapshot();
    });
    it('with params', () => {
      expect(
        new AliOSSImageProcessingUtil(
          `${originUrl}?param1=1&param2=a`
        ).getAliOSSImageAutoOrientAPI({
          autoOrient: false,
        })
      ).toMatchSnapshot();
      expect(
        new AliOSSImageProcessingUtil(
          `${originUrl}?param1=1&param2=a&`
        ).getAliOSSImageAutoOrientAPI({
          autoOrient: false,
        })
      ).toMatchSnapshot();
    });
  });
});
