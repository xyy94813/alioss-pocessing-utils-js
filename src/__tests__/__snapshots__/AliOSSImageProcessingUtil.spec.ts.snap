// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AliOSSImageProcessingUtil getAliOSSImageAutoOrientAPI auto origent 1`] = `"https://image-demo.oss-cn-hangzhou.aliyuncs.com/example.jpg?x-oss-process=image%2Fauto-orient%2C1"`;

exports[`AliOSSImageProcessingUtil getAliOSSImageAutoOrientAPI not auto origent 1`] = `"https://image-demo.oss-cn-hangzhou.aliyuncs.com/example.jpg?x-oss-process=image%2Fauto-orient%2C0"`;

exports[`AliOSSImageProcessingUtil getAliOSSImageCircleAPI radius is normal 1`] = `"https://image-demo.oss-cn-hangzhou.aliyuncs.com/example.jpg?x-oss-process=image%2Fcircle%2Cr_100"`;

exports[`AliOSSImageProcessingUtil getAliOSSImageCropAPI set width and height 1`] = `"https://image-demo.oss-cn-hangzhou.aliyuncs.com/example.jpg?x-oss-process=image%2Fcrop%2Ch_100%2Cw_100"`;

exports[`AliOSSImageProcessingUtil getAliOSSImageIndexcropAPI x is 100 and i is 0 1`] = `"https://image-demo.oss-cn-hangzhou.aliyuncs.com/example.jpg?x-oss-process=image%2Findexcrop%2Ci_100%2Cx_100"`;

exports[`AliOSSImageProcessingUtil getAliOSSImageProcessingAPI mutiple operation 1`] = `"https://image-demo.oss-cn-hangzhou.aliyuncs.com/example.jpg?x-oss-process=image%2Fresize%2Ch_200%2Cw_200%2Frotate%2C90"`;

exports[`AliOSSImageProcessingUtil getAliOSSImageResizeAPI with mode mode is fill 1`] = `"https://image-demo.oss-cn-hangzhou.aliyuncs.com/example.jpg?x-oss-process=image%2Fresize%2Cm_fill%2Cw_200"`;

exports[`AliOSSImageProcessingUtil getAliOSSImageResizeAPI with mode mode is fixed 1`] = `"https://image-demo.oss-cn-hangzhou.aliyuncs.com/example.jpg?x-oss-process=image%2Fresize%2Cm_fixed%2Cw_200"`;

exports[`AliOSSImageProcessingUtil getAliOSSImageResizeAPI with mode mode is lfit 1`] = `"https://image-demo.oss-cn-hangzhou.aliyuncs.com/example.jpg?x-oss-process=image%2Fresize%2Cm_lfit%2Cw_200"`;

exports[`AliOSSImageProcessingUtil getAliOSSImageResizeAPI with mode mode is mfit 1`] = `"https://image-demo.oss-cn-hangzhou.aliyuncs.com/example.jpg?x-oss-process=image%2Fresize%2Cm_mfit%2Cw_200"`;

exports[`AliOSSImageProcessingUtil getAliOSSImageResizeAPI with mode mode is pad 1`] = `"https://image-demo.oss-cn-hangzhou.aliyuncs.com/example.jpg?x-oss-process=image%2Fresize%2Cm_pad%2Cw_200"`;

exports[`AliOSSImageProcessingUtil getAliOSSImageRotateAPI set rotate 1`] = `"https://image-demo.oss-cn-hangzhou.aliyuncs.com/example.jpg?x-oss-process=image%2Frotate%2C30"`;

exports[`AliOSSImageProcessingUtil getAliOSSImageRoundedCornersAPI set radius 1`] = `"https://image-demo.oss-cn-hangzhou.aliyuncs.com/example.jpg?x-oss-process=image%2Frounded-corners%2Cr_30"`;

exports[`AliOSSImageProcessingUtil origin url with query with params 1`] = `"https://image-demo.oss-cn-hangzhou.aliyuncs.com/example.jpg?param1=1&param2=a&x-oss-process=image%2Fauto-orient%2C0"`;

exports[`AliOSSImageProcessingUtil origin url with query with params 2`] = `"https://image-demo.oss-cn-hangzhou.aliyuncs.com/example.jpg?param1=1&param2=a&x-oss-process=image%2Fauto-orient%2C0"`;

exports[`AliOSSImageProcessingUtil origin url with query without param 1`] = `"https://image-demo.oss-cn-hangzhou.aliyuncs.com/example.jpg?x-oss-process=image%2Fauto-orient%2C0"`;
